New gadget: fsnotify

alban opened the pull request:
# New gadget: fsnotify

Fixes: https://github.com/inspektor-gadget/inspektor-gadget/issues/2903

## How to use

See gadgets/fsnotify/README.md

## Testing done

- Only compiled the new gadget

burak-ok commented on gadgets/fsnotify/artifacthub-pkg.yml:
```diff
@@ -0,0 +1,29 @@
+# Artifact Hub package metadata file
+version: 0.1.0
```
```suggestion
version: 0.31.0
```

burak-ok reviewed (commented):
A lot have changed in the last week in regarding to the docs. I didn't comment everything in these files

For `artifacthub-pkg.yml`, `gadget.yaml`, `README.md` and `README.mdx` could you check / copy from (for example) `trace_exec` these files and change them to your needs.

Also a symbolic link from `docs/gadgets/fsnotify.mdx` to `gadgets/fsnotify/README.mdx` should be made

alban commented:
`--update-metadata` lost the metadata for all fields (documentation, attributes, templates) because the fields were moved from `structs` to `datasources`. If such a change in gadget.yaml happens again, we should provide tooling for users to update their gadgets because it is a lot of manual work to update.

alban commented on gadgets/fsnotify/README.md:
```diff
@@ -0,0 +1,5 @@
+
+The `fsnotify` gadget detects applications using inotify or fanotify.
+
+Check the full documentation on https://inspektor-gadget.io/docs/latest/gadgets
```
I removed the /fsnotify suffix because the CI fails.
How to resolve this chicken egg problem?

mauriciovasquezbernal replied:
I don't see we have any solution there. I'd propose to ignore the CI and use the right link there.

alban commented on gadgets/fsnotify/artifacthub-pkg.yml:
```diff
@@ -0,0 +1,29 @@
+description: "enrich inotify or fanotify events"
+logoURL: "https://inspektor-gadget.io/media/brand-icon.svg"
+license: ""
+homeURL: "https://inspektor-gadget.io/docs/latest/gadgets"
```
I removed the /fsnotify suffix because the link is not yet valid.
How to resolve this chicken egg problem?

mauriciovasquezbernal replied:
I think it's better to use the right link. We should have an automatic way to update the website when changes are done to the docs folder here. I'll think about it.

mauriciovasquezbernal replied:
For reference, it was done in #3342.

alban commented:
Updated. The only broken link revealed by the CI is https://inspektor-gadget.io/docs/latest/gadgets/fsnotify

burak-ok commented:
I thought I had posted the comment in my previous review, but I can't find it anymore. Nonetheless I have the following question:

Do we want to call this gadget just `fsnotify`. Isn't this more of a `trace_fsnotify`? This would also align it with all other gadgets

mauriciovasquezbernal reviewed (commented):
Some comments. I haven't gone through the whole ebpf code, but please don't block this on me.

alban commented on gadgets/fsnotify/README.mdx:
````diff
@@ -0,0 +1,118 @@
+        ```
+    </TabItem>
+</Tabs>
+## Flags
+
+### `--fanotify-only`
+Show only fanotify events
+
+Default value: ""
+
+### `--inotify-only`
+Show only inotify events
+
+Default value: ""
+
+### `--tracee-pid`
+Show only fsnotify events generated by processes with this pid
+
+Default value: ""
+
+### `--tracer-group`
+Show only fsnotify events generated by a struct fsnotify_group with this address. You can find this address in Golang with kfilefields.ReadPrivateDataFromFd().
+
+Default value: ""
+
+### `--tracer-pid`
+Show only fsnotify events generated for processes with this pid
+
+Default value: ""
````
I can add the spaces. But it comes from the template, so ideally we should fix it in the template (#3357).

alban commented on gadgets/fsnotify/README.mdx:
````diff
@@ -0,0 +1,118 @@
+
+The fsnotify-enricher gadget can observe and enrich the inotify events in the following way:
+```
+$ sudo -E ig run ghcr.io/inspektor-gadget/gadget/fsnotify:latest --verify-image=false --inotify-only --fields=tracer_comm,tracee_comm,i_wd,i_mask,i_cookie,name
````
> -E is not needed anymore :)

Ok, I will remove it.

But in general it is needed to get `$HOME` to get the default docker auth file. In this case, `ghcr.io/inspektor-gadget/gadget/fsnotify:latest` is a public image, but without `-E` won't work for private images.

```bash
$ sudo sh -c 'echo $HOME'
/root
$ sudo -E sh -c 'echo $HOME'
/home/alban
```

alban commented:
Thanks for the reviews. Updated.

About the naming: since this gadget does not come from a built-in gadget, we don't have the need to keep the category, so "fsnotify" is fine.

alban commented:
![gadget-fsnotify-architecture](https://github.com/user-attachments/assets/daf8e4c9-4cb7-4e38-aae5-bef7a6b316d6)

alban commented:
The diagram above explains most of the architecture but some concepts couldn't be represented on it:
- merged events
- filters, particularly how the gadget can be used in combination with `kfilefields.ReadPrivateDataFromFd`
- how fanotify permission responses are sent
- kprobe/fsnotify_destroy_event is called from the tracee in case of merged event or fanotify permission response event, but from the tracer in other cases.

mauriciovasquezbernal commented on gadgets/fsnotify/artifacthub-pkg.yml:
```diff
@@ -0,0 +1,29 @@
+    - gadget
+links:
+    - name: source
+      url: "https://github.com/inspektor-gadget/inspektor-gadget/"
```
point this to the exact location of the gadget?

mauriciovasquezbernal commented on gadgets/fsnotify/gadget.yaml:
```diff
@@ -0,0 +1,117 @@
+name: fsnotify
+description: Gadget enriching inotify or fanotify events
+homepageURL: https://inspektor-gadget.io/
+documentationURL: https://inspektor-gadget.io/docs
+sourceURL: https://github.com/inspektor-gadget/inspektor-gadget/
```
ah, not related to this PR, but should probably update all gadgets to point to the specific gadget docs and source.

alban replied:
Done for other gadgets in #3376

mauriciovasquezbernal commented on gadgets/fsnotify/gadget.yaml:
```diff
@@ -0,0 +1,117 @@
+      fa_pid:
+        annotations:
+          description: pid from fanotify_event_metadata. Process ID, or thread ID
+            if flag FAN_REPORT_TID was set in fanotify_init
```
template: pid?

mauriciovasquezbernal commented on gadgets/fsnotify/gadget.yaml:
```diff
@@ -0,0 +1,117 @@
+            if flag FAN_REPORT_TID was set in fanotify_init
+      fa_response:
+        annotations:
+          description: 'response to a fanotify permission request: allow or deny'
```
nit: Why `'` are used in some cases and no in others?

alban replied:
When the string contains a `:`, it needs to be quoted, otherwise the yaml parser gets confused.
I think that `--update-metadata` removes the quotes when not needed.
About `'` vs `"`, I don't know.

mauriciovasquezbernal replied:
You got a point!, I didn't realize there were too many `:` in this file.

mauriciovasquezbernal commented on gadgets/fsnotify/program.bpf.c:
```diff
@@ -0,0 +1,523 @@
+GADGET_PARAM(inotify_only);
+GADGET_PARAM(fanotify_only);
+
+struct gadget_event {
```
Which one from mntns_id or tracee_mntns_id should be used for enrichment. Currently we only support one of them per struct. 

```bash 
WARN[0000] multiple fields of "type:gadget_mntns_id" found in DataSource "fsnotify", only "mntns_id" will be used
```

mauriciovasquezbernal replied:
Please either add `_raw` suffix or use the `ebpf.formatter.enum` annotation for enumerations to work: https://www.inspektor-gadget.io/docs/latest/gadget-devel/gadget-ebpf-api#enumerations

```bash
WARN[0000] Failed to get target name for enum field "type": neither "ebpf.formatter.enum" annotation nor '_raw' suffix found
WARN[0000] Failed to get target name for enum field "fa_type": neither "ebpf.formatter.enum" annotation nor '_raw' suffix found
WARN[0000] Failed to get target name for enum field "fa_response": neither "ebpf.formatter.enum" annotation nor '_raw' suffix found
```

mauriciovasquezbernal commented on gadgets/fsnotify/README.mdx:
````diff
@@ -0,0 +1,117 @@
+```bash
+$ sudo ig run ghcr.io/inspektor-gadget/gadget/fsnotify:latest --inotify-only --fields=tracer_comm,tracee_comm,i_wd,i_mask,i_cookie,name
+TRACER_COMM  TRACEE_COMM I_WD I_MASK    I_COOKIE NAME
+inotifywatch touch       1    134217760 0        ABCDE
+inotifywatch touch       1    134217732 0        ABCDE
+inotifywatch touch       1    134217736 0        ABCDE
````
Is is expected for NAME not to include the whole path?

alban replied:
It's not implemented yet, see: https://github.com/alban/fsnotify-enricher/issues/1

It is not that easy to implement because at the time the kprobes are called, the kernel data structures only contain the name string and the inode number (in the inotify case) but no `struct file` or `struct inode`.

mauriciovasquezbernal commented on gadgets/fsnotify/README.mdx:
Please add the link to docs/gadgets too.

mauriciovasquezbernal reviewed (commented):
Some general comments. I tested as indicated in the guide and it works fine. Unfortunately I don't have a deep understanding of the fanotify/inotify mechanisms to provide a better review, however what I reviewed in the code makes sense. 

I'll try to make another pass once the these nits are fixed.

alban commented:
I think I addressed all the review comments.

After #3284 is completed, I'd like to update the guide to make use of it.

After #3132 is completed, I would like parameters (`--inotify-only`, `--fanotify-only`) to automatically hide/delete some columns (hide fanotify specific columns when `--inotify-only` is specified).

alban commented:
Merged by squashing commits 1 and 2 as suggested.
CI tests pass (except the URL check on the documentation as expected).
